pontoon/sync/formats/json_extensions.py

```diff
@@ -0,0 +1,179 @@
+    """
+    Represents an entity in a JSON file.
+    """
+    def __init__(self, order, key, data):
+        self.key = key
+        self.data = data
+        self.order = order
+        self.strings = {None: self.source_string} if self.source_string else {}
+        self.last_translator = None
+        self.last_update = None
+
+    @property
+    def source_string(self):
+        return self.data['message']
+
+    @property
+    def source_string_plural(self):
+        return ''
+
+    @property
+    def comments(self):
+        return [self.data['description']] if 'description' in self.data else []
+
+    @property
+    def fuzzy(self):
+        return False
+
+    @fuzzy.setter
+    def fuzzy(self, fuzzy):
+        pass  # We don't use fuzzy in JSON
+
+    @property
+    def source(self):
+        return self.data.get('placeholders', [])
+
+
+class JSONResource(ParsedResource):
+    def __init__(self, path, source_resource=None):
+        self.path = path
+        self.entities = {}
+        self.source_resource = source_resource
+
+        # Copy entities from the source_resource if it's available.
+        if source_resource:
+            for key, entity in source_resource.entities.items():
+                self.entities[key] = JSONEntity(
+                    entity.order,
+                    entity.key,
+                    copy.copy(entity.data),
+                )
+
+        try:
+            with codecs.open(path, 'r', 'utf-8') as resource:
+                self.json_file = json.load(resource, object_pairs_hook=OrderedDict)
+                validate(self.json_file, SCHEMA)
+
+        except (IOError, ValueError, ValidationError) as err:
+            # If the file doesn't exist or cannot be decoded,
+            # but we have a source resource,
+            # we can keep going, we'll just not have any translations.
+            if source_resource:
+                return
+            else:
+                raise ParseError(err)
+
+        for order, (key, data) in enumerate(self.json_file.items()):
+            self.entities[key] = JSONEntity(
+                order,
+                key,
+                data,
+            )
+
+    @property
+    def translations(self):
+        return sorted(self.entities.values(), key=lambda e: e.order)
+
+    def save(self, locale):
+        """
+        Load the source resource, modify it with changes made to this
+        Resource instance, and save it over the locale-specific
+        resource.
+        """
+        if not self.source_resource:
+            raise SyncError('Cannot save FTL resource {0}: No source resource given.'
+                            .format(self.path))
+
+        with codecs.open(self.source_resource.path, 'r', 'utf-8') as resource:
+            json_file = json.load(resource, object_pairs_hook=OrderedDict)
+
+            try:
+                validate(json_file, SCHEMA)
+            except ValidationError as e:
+                raise ParseError(e)
```

Comment:
Should we really refuse to sync if a translated json isn't valid? Or should we rather fix it up?